full-service/src/service/transaction_builder.rs

```diff
@@ -430,20 +446,43 @@ impl<FPR: FogPubkeyResolver + 'static> WalletTransactionBuilder<FPR> {
             }
 
             let change_amount = Amount::new(change_value as u64, token_id);
-            let tx_out_context = transaction_builder.add_change_output(
-                change_amount,
-                &reserved_subaddresses,
-                &mut rng,
-            )?;
+            if let Some(spend_from_subaddress) = self.spend_only_from_subaddress.as_ref() {
+                // Send the change back to the subaddress that is spending the inputs.
+                // In the future, we may want to allow this to be a bit more configurable
+                let change_address = b58_decode_public_address(&spend_from_subaddress)?;
```

Comment:
** ⚠️  IMPORTANT IMPLEMENTATION DETAIL ⚠️ **

I made the decision to send the change back to the subaddress from which the transaction is being spent. 🚨 THIS BREAKS RECOVERABLE TRANSACTION HISTORY.

**Why do I think this is ok?**

* By choosing to spend only from a given subaddress, the full-service subaddress model is being broken. The subaddresses are meant to be turnstiles, so that you can track funds coming in, but you can spend from any txo. This is important to several design decisions made for this wallet, and is meant to lend to efficiency and simplicity of logic.
* This new model with spending only from subaddresses implies that a subaddress is more like a subaccount. The challenge here is that an ACCOUNT only has ONE change address, the default change address.

The options for where change could go:

1. Change goes to the default change subaddress. ❌ Then the balance for the subaddress would instantly be wrong, because typically there is only one TXO which is then spent from, and that change would no longer be associated with the spending subaddress.
2. Change goes back to the subaddress from which it was spent. 🟢 This is ok, BUT, there is no way currently to write a Destination memo for any output not going to the default change subaddress.
3. Create a new scheme where each subaddress has an associated change subaddress, and we increment by two subaddresses whenever a subaddress is created. ❌ This is more complicated than necessary at this point. If we find that people really want recoverable transaction history AND subaddresses that are actually behaving like subaccounts, we could revisit this. As it stands, the transaction history is preserved with transaction logs.

Reply:
What technically precludes including a destination memo on the change txo when it goes back to the subaddress from which the spend is happening?

Does this trait implementation not allow overriding the change subaddress in a way that would cause the destination memo to be properly written? https://github.com/mobilecoinfoundation/mobilecoin/blob/39cdbb21b89e173ee7b3f1666a7f65b8ff7e58e9/transaction/builder/src/reserved_subaddresses.rs#L57

There is, and should be noted, a reason to distrust destination memos sent by anybody but the originator of the change, so a wallet implementer would want to evaluate the threat model and ramifications of possibly forged destination memos before deciding to make use of them.

Reply:
Wow 🌮 That is exactly what I needed! So much emoji angst above 😅 when the API was already available to do exactly what we needed! Code is updated, and I'm going to clean up the builder call a little bit, but please take a look!

Thank you @holtzman!